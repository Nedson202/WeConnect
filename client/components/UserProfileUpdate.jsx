import React, { Component } from 'react';
import classcat from 'classcat';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import ImageUpload from './ImageUpload';
import ImageUploader from '../actions/imageUpload';
import {addFlashMessage} from '../actions/flashMessages';
import '../index.scss';

/**
 * @class UserProfileUpdate
 * 
 * @extends {Component}
 */
class UserProfileUpdate extends Component {
  /**
   * @description Creates an instance of UserProfileUpdate.
   * 
   * @param {object} props 
   * 
   * @memberof UserProfileUpdate
   */
  constructor(props){
    super(props);

    this.state = {
      username: this.props.user ? this.props.user.username : "",
      email: this.props.user ? this.props.user.email : "",
      errors: {}
    };

    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  /**
   * @description update state with data from input
   * 
   * @param {any} event
   * 
   * @returns {undefined}
   * 
   * @memberof UserProfileUpdate
   */
  onChange(event) {
    if (this.state.errors[event.target.name]) {
      let errors = Object.assign({}, this.state.errors);
      delete errors[event.target.name];
      this.setState({
        [event.target.name]: event.target.value.toLowerCase(),
        errors
      });
    }
    this.setState({ [event.target.name]: event.target.value });
  }
  /**
   * @description handle submission of profile
   * 
   * @param {any} event
   * 
   * @returns {undefined}
   * 
   * @memberof UserProfileUpdate
   */
  onSubmit(event) {
    event.preventDefault();    
    this.props.userProfileUpdateRequest(this.props.user.userId, this.state).then(
      () => {
        document.getElementById('close').click();        
      }
    )
  }

  /**
   * @description Renders the component to the dom
   * 
   * @returns {object} JSX object
   * 
   * @memberof UserProfileUpdate
   */
  render() {
    const { errors } = this.state;

    return (
      <div>
        <div className="modal fade" id="updateModal" tabIndex="-1" role="dialog" aria-labelledby="confirmModalTitle" aria-hidden="true">
          <div className="modal-dialog modal-dialog-centered" role="document">
            <div className="modal-content">
              <div className="modal-header">
                <h4>Edit profile</h4>
                <button type="button" className="close" data-dismiss="modal" aria-label="Close" id="close">
                  <span aria-hidden="true">&times;</span>
                </button>
              </div>
              <div className="modal-body">
                <form>
                  <div className="row"> 
                    <div className={classcat(["form-group",
                        { "has-error": errors.username },
                        'col-lg-12'
                      ])}
                    >
                      <label  id="control-label">Username</label>
                      <input
                        value={this.state.username}
                        onChange={this.onChange}
                        type="text"
                        name="username"
                        className="form-control"
                        id="control-label"
                        placeholder="username"
                      />
                      {errors && <span className="help-block">{errors.username}</span>}
                    </div>
                    <div className={classcat(["form-group",
                        { "has-error": errors.email },
                        'col-lg-12'
                      ])}
                    >
                      <label  id="control-label">Email</label>
                      <input
                        value={this.state.email}
                        onChange={this.onChange}
                        type="text"
                        name="email"
                        className="form-control"
                        id="control-label"
                        placeholder="email"
                      />
                      {errors && <span className="help-block">{errors.email}</span>}
                    </div>
                    <ImageUpload 
                      ImageUploader={this.props.ImageUploader} 
                      user={this.props.user} 
                      addFlashMessage={this.props.addFlashMessage}
                    />      
                  </div>
                </form>
              </div>
              <div className="modal-footer">
                <button type="button" className="btn btn-outline-success" data-dismiss="modal">No</button>
                <button 
                  type="button" 
                  className="btn btn-outline-success" 
                  onClick={this.onSubmit} 
                  data-dismiss="modal"
                >Yes
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

UserProfileUpdate.propTypes = {
  ImageUploader: PropTypes.func.isRequired,
  addFlashMessage: PropTypes.func.isRequired
}

const mapDispatchToProps = dispatch => bindActionCreators({
  ImageUploader,
  addFlashMessage
}, dispatch)

export default connect(null, mapDispatchToProps)(UserProfileUpdate);