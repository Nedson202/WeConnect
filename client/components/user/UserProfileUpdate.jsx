import React from 'react';
import PropTypes from 'prop-types';
import ImageUpload from '../ImageUpload/ImageUpload';

/**
 * @class UserProfileUpdate
 *
 * @extends {Component}
 */
const UserProfileUpdate = ({
  state, onChange, onSubmit, user, uploadToCloudinary, onImageChange
}) => {
  const { username, email } = state;
  return (
    <div>
      <div className="modal fade" id="updateModal" tabIndex="-1" role="dialog" aria-labelledby="confirmModalTitle" aria-hidden="true">
        <div className="modal-dialog modal-dialog-centered" role="document">
          <div className="modal-content">
            <div className="modal-header">
              <h4>Edit profile</h4>
              <button type="button" className="close" data-dismiss="modal" aria-label="Close" id="closeProfileModal">
                <span aria-hidden="true">&times;</span>
              </button>
            </div>
            <div className="modal-body">
              <form>
                <div className="row">
                  <div className="form-group col-lg-12">
                    <label id="upload-form-label">Username</label>
                    <input
                      value={username}
                      onChange={onChange}
                      type="text"
                      name="username"
                      className="form-control"
                      id="control-label"
                      placeholder="username"
                    />
                  </div>
                  <div className="form-group col-lg-12">
                    <label id="upload-form-label">Email</label>
                    <input
                      value={email}
                      onChange={onChange}
                      type="text"
                      name="email"
                      className="form-control"
                      id="control-label"
                      placeholder="email"
                    />
                  </div>
                  <ImageUpload
                    onImageChange={onImageChange}
                    uploadToCloudinary={uploadToCloudinary}
                    user={user}
                    state={state}
                  />
                </div>
              </form>
            </div>
            <div className="modal-footer">
              <button type="button" className="btn btn-outline-success" data-dismiss="modal">No</button>
              <button
                className="btn btn-outline-success"
                onClick={onSubmit}
                id="profile-update-button"
              >Upload
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

UserProfileUpdate.propTypes = {
  onChange: PropTypes.func.isRequired,
  onImageChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  state: PropTypes.object.isRequired,
  uploadToCloudinary: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired,
};

export default UserProfileUpdate;
