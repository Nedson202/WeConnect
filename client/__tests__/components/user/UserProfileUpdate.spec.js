import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJSON from 'enzyme-to-json';
import UserProfileUpdate from '../../../components/user/UserProfileUpdate';
import userInfo from '../../__mockData__/userData';

configure({ adapter: new Adapter() });
const { state, userData } = userInfo;
let props;
const setup = () => {
  props = {
    state,
    onChange: jest.fn(),
    onSubmit: jest.fn(),
    user: userData,
    uploadToCloudinary: jest.fn(),
    onImageChange: jest.fn()
  };

  return shallow(<UserProfileUpdate {...props} />);
};


describe('Component: UserProfileUpdate', () => {
  it('should render', () => {
    const wrapper = setup();
    expect(wrapper.find('div').length).toBe(10);
    expect(toJSON(wrapper)).toMatchSnapshot();
    expect(wrapper.find('h4').length).toBe(1);
    expect(wrapper.find('h4').text()).toBe('Edit profile');
    expect(wrapper.find('form').length).toBe(1);
    expect(wrapper.find('.close').simulate('click'));
    expect(wrapper.find('#profile-update-button').simulate('click'));
    expect(wrapper.length).toBe(1);
  }) 
});