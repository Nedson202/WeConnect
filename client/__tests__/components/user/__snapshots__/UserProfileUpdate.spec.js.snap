// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: UserProfileUpdate should render 1`] = `
<div>
  <div
    aria-hidden="true"
    aria-labelledby="confirmModalTitle"
    className="modal fade"
    id="updateModal"
    role="dialog"
    tabIndex="-1"
  >
    <div
      className="modal-dialog modal-dialog-centered"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h4>
            Edit profile
          </h4>
          <button
            aria-label="Close"
            className="close"
            data-dismiss="modal"
            id="closeProfileModal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          <form>
            <div
              className="row"
            >
              <div
                className="form-group col-lg-12"
              >
                <label
                  id="upload-form-label"
                >
                  Username
                </label>
                <input
                  className="form-control"
                  id="control-label"
                  name="username"
                  onChange={[MockFunction]}
                  placeholder="username"
                  type="text"
                  value="mike"
                />
              </div>
              <div
                className="form-group col-lg-12"
              >
                <label
                  id="upload-form-label"
                >
                  Email
                </label>
                <input
                  className="form-control"
                  id="control-label"
                  name="email"
                  onChange={[MockFunction]}
                  placeholder="email"
                  type="text"
                  value="alen@stuart.mill"
                />
              </div>
              <ImageUploader
                onImageChange={[MockFunction]}
                state={
                  Object {
                    "email": "alen@stuart.mill",
                    "username": "mike",
                  }
                }
                uploadToCloudinary={[MockFunction]}
                user={
                  Object {
                    "id": 1,
                    "username": "astro",
                  }
                }
              />
            </div>
          </form>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-outline-success"
            data-dismiss="modal"
            type="button"
          >
            No
          </button>
          <button
            className="btn btn-outline-success"
            id="profile-update-button"
            onClick={[MockFunction]}
          >
            Upload
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
